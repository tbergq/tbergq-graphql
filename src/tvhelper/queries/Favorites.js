// @flow

import {
  fromGlobalId,
  connectionFromArray,
  connectionArgs,
  type ConnectionArguments,
} from 'graphql-relay';

import TvShowConnection from '../types/TvShowConnection';
import type { GraphqlContextType } from '../../common/services/GraphqlContext';
import type { TvShow as TvShowType } from '../dataloaders/SearchTvShowLoader';

export default {
  name: 'Favorites',
  type: TvShowConnection,
  description: 'Get your favorites',
  args: {
    ...connectionArgs,
  },
  resolve: async (
    _: mixed,
    args: ConnectionArguments,
    { user, dataLoader }: GraphqlContextType,
  ) => {
    console.log(user);
    if (user == null) {
      throw Error('You must be signed in to use this query');
    }
    const { id } = fromGlobalId(user.id);
    const savedFavorites = await dataLoader.tvhelper.favorites.load(id);
    const serieIds = savedFavorites.map(item => item.serieId.toString());
    const favorites = await dataLoader.tvhelper.tvDetail.loadMany(serieIds);

    return connectionFromArray<TvShowType>(favorites, args);
  },
};
