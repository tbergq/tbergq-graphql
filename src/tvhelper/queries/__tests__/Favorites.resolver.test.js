// @flow

import Favorites from '../Favorites';

const { resolve } = Favorites;

const ancestor = null;
const user = {
  id: '1',
};
const dataLoader = {
  tvhelper: {
    favorites: {
      load: async () =>
        Promise.resolve([
          { serieId: 1 },
          { serieId: 2 },
          { serieId: 3 },
          { serieId: 4 },
        ]),
    },
    tvDetail: {
      loadMany: async () =>
        Promise.resolve([
          {
            name: 'Game of Thrones',
            status: 'running',
            _embedded: {
              nextepisode: { airdate: new Date('2018-01-01') },
              previousepisode: { airdate: new Date('2018-01-01') },
            },
          },
          {
            name: 'Animal kingdom',
            status: 'ended',
            _embedded: {
              nextepisode: { airdate: new Date('2018-12-01') },
              previousepisode: { airdate: new Date('2018-12-01') },
            },
          },
          {
            name: 'The 100',
            status: 'in development',
            _embedded: {
              nextepisode: { airdate: new Date('2018-12-05') },
              previousepisode: { airdate: new Date('2018-12-05') },
            },
          },
          {
            name: 'Elite',
            status: 'running',
            _embedded: {
              nextepisode: { airdate: new Date('2017-12-01') },
              previousepisode: { airdate: new Date('2017-12-01') },
            },
          },
        ]),
    },
  },
};
const context = {
  user,
  dataLoader,
};

const ascending = 'ascending';
const descending = 'descending';

it('sorts by name ascending', async () => {
  expect(
    await resolve(
      ancestor,
      { options: { sortDirection: ascending, sortBy: 'name' } },
      // $FlowExpectedError: Passing just dataLoaders needed to test
      context,
    ),
  ).toMatchSnapshot();
});

it('sorts by name descending', async () => {
  expect(
    await resolve(
      ancestor,
      { options: { sortDirection: descending, sortBy: 'name' } },
      context,
    ),
  ).toMatchSnapshot();
});

it('sorts by status ascending', async () => {
  expect(
    await resolve(
      ancestor,
      { options: { sortDirection: ascending, sortBy: 'status' } },
      context,
    ),
  ).toMatchSnapshot();
});

it('sorts by status descending', async () => {
  expect(
    await resolve(
      ancestor,
      { options: { sortDirection: descending, sortBy: 'status' } },
      context,
    ),
  ).toMatchSnapshot();
});

it('sorts by nextepisode ascending', async () => {
  expect(
    await resolve(
      ancestor,
      {
        options: {
          sortDirection: ascending,
          sortBy: '_embedded.nextepisode.airdate',
        },
      },
      context,
    ),
  ).toMatchSnapshot();
});

it('sorts by nextepisode descending', async () => {
  expect(
    await resolve(
      ancestor,
      {
        options: {
          sortDirection: descending,
          sortBy: '_embedded.nextepisode.airdate',
        },
      },
      context,
    ),
  ).toMatchSnapshot();
});

it('sorts by previousepisode ascending', async () => {
  expect(
    await resolve(
      ancestor,
      {
        options: {
          sortDirection: ascending,
          sortBy: '_embedded.previousepisode.airdate',
        },
      },
      context,
    ),
  ).toMatchSnapshot();
});

it('sorts by previousepisode descending', async () => {
  expect(
    await resolve(
      ancestor,
      {
        options: {
          sortDirection: descending,
          sortBy: '_embedded.previousepisode.airdate',
        },
      },
      context,
    ),
  ).toMatchSnapshot();
});
