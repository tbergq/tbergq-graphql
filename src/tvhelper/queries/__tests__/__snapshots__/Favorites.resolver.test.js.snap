// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sorts by name ascending 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
        },
        "name": "Animal kingdom",
        "status": "ended",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
        },
        "name": "Elite",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
        },
        "name": "Game of Thrones",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
        },
        "name": "The 100",
        "status": "in development",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`sorts by name descending 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
        },
        "name": "The 100",
        "status": "in development",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
        },
        "name": "Game of Thrones",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
        },
        "name": "Elite",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
        },
        "name": "Animal kingdom",
        "status": "ended",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`sorts by nextepisode ascending 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
        },
        "name": "Elite",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
        },
        "name": "Game of Thrones",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
        },
        "name": "Animal kingdom",
        "status": "ended",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
        },
        "name": "The 100",
        "status": "in development",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`sorts by nextepisode descending 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
        },
        "name": "The 100",
        "status": "in development",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
        },
        "name": "Animal kingdom",
        "status": "ended",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
        },
        "name": "Game of Thrones",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
        },
        "name": "Elite",
        "status": "running",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`sorts by previousepisode ascending 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
        },
        "name": "Elite",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
        },
        "name": "Game of Thrones",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
        },
        "name": "Animal kingdom",
        "status": "ended",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
        },
        "name": "The 100",
        "status": "in development",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`sorts by previousepisode descending 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
        },
        "name": "The 100",
        "status": "in development",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
        },
        "name": "Animal kingdom",
        "status": "ended",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
        },
        "name": "Game of Thrones",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
        },
        "name": "Elite",
        "status": "running",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`sorts by status ascending 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
        },
        "name": "Animal kingdom",
        "status": "ended",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
        },
        "name": "The 100",
        "status": "in development",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
        },
        "name": "Game of Thrones",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
        },
        "name": "Elite",
        "status": "running",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;

exports[`sorts by status descending 1`] = `
Object {
  "edges": Array [
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-01-01T00:00:00.000Z,
          },
        },
        "name": "Game of Thrones",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2017-12-01T00:00:00.000Z,
          },
        },
        "name": "Elite",
        "status": "running",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-05T00:00:00.000Z,
          },
        },
        "name": "The 100",
        "status": "in development",
      },
    },
    Object {
      "cursor": "YXJyYXljb25uZWN0aW9uOjM=",
      "node": Object {
        "_embedded": Object {
          "nextepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
          "previousepisode": Object {
            "airdate": 2018-12-01T00:00:00.000Z,
          },
        },
        "name": "Animal kingdom",
        "status": "ended",
      },
    },
  ],
  "pageInfo": Object {
    "endCursor": "YXJyYXljb25uZWN0aW9uOjM=",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "YXJyYXljb25uZWN0aW9uOjA=",
  },
}
`;
